File: packages/mcp-server/src/lib/tokens.ts
/**
 * Token estimation.
 *
 * The token budget is the agent's scarcest resource (Part I.1). Every retrieval
 * respects it. The estimator is pluggable so we can later swap the cheap
 * heuristic for a real tokenizer without touching call sites.
 */
export interface TokenEstimator {
  estimate(text: string): number;
}

/** Cheap heuristic: ~4 characters per token. Good enough for budgeting. */
export const heuristicEstimator: TokenEstimator = {
  estimate(text: string): number {
    if (!text) return 0;
    return Math.ceil(text.length / 4);
  },
};

export function estimate(text: string, estimator: TokenEstimator = heuristicEstimator): number {
  return estimator.estimate(text);
}

File: crates/tacitus-core/src/tokens.rs
/// Token estimation — the agent's scarcest resource is the context window.
/// Cheap heuristic: ~4 characters per token (mirrors the TS `heuristicEstimator`).
pub fn estimate(text: &str) -> usize {
    let chars = text.chars().count();
    chars.div_ceil(4)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn estimates_ceil_of_len_over_four() {
        assert_eq!(estimate(""), 0);
        assert_eq!(estimate("abcd"), 1);
        assert_eq!(estimate("abcde"), 2);
    }
}
